Enable database fetchers for discovery service

greedy52 opened the pull request:
Implements
- #18742 

Part 1
- #19510

This change is a redo of #19030 after the above refactoring.


### UX
Discovery service uses the same matchers from database service today, e.g:
```
discovery_service:
  enabled: "yes"
  aws:
  - types: ["redshift", "rds"]
    regions: ["ca-central-1"]
    tags:
      "Owner": "STeve"

  - types: ["eks"]
    regions: ["us-west-2"]
    tags:
      "*": "*"

  azure:
  - types: ["redis"]
    resource_groups: ["STeve-teleport"]
    tags:
      "*": "*"
```

Database service filters databases created by discovery services through **existing** resource labels:
```
db_service:
  enabled: "yes"
  resources:
    - labels:
        "*": "*"
```
Database service (at this point) can still use existing aws/azure matchers to discover databases by itself for backward compatibility.

greedy52 committed:
Enable database fetchers for discovery service

greedy52 committed:
add missing redshift serverless type

greedy52 commented:
@r0mant @smallinsky could you review if the **UX** section in the description is OK? 

There were some discussions about how should the database agent handles the existing matchers and the databases created from discovery services. I implemented the same logic as the kube service as I think it's more straight forward this way but there are concerns for example existing database agents with wildcard resource matchers will gain everything from discovery service.

smallinsky commented:
@greedy52 
I think that this approach is reasonable to reuse the database dynamic `"*": "*"` resource matcher. We can consider checking if db agent that suppose to handle incoming cloud db database discovery resource can actually successfully handle them by ensuring that db agent runs with AWS account by calling  [API_GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html)   in case of Azure we have similar API like `az account show`. Additionally we can check the AccountID/subscriptionID of an incoming resource and in case of multi AWS/Azure cloud configuration.

greedy52 commented:
> @greedy52 I think that this approach is reasonable to reuse the database dynamic `"*": "*"` resource matcher. We can consider checking if db agent that suppose to handle incoming cloud db database discovery resource can actually successfully handle them by ensuring that db agent runs with AWS account by calling [API_GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html) in case of Azure we have similar API like `az account show`. Additionally we can check the AccountID/subscriptionID of an incoming resource and in case of multi AWS/Azure cloud configuration.

I totally understand the concern. Though in my opinion, I believe this level of filtering is best manually added to the dynamic resource labels. It is possible for a database agent to access databases in a different account (e.g. assume role in a different account for DynamoDB), and it is also possible for a database agent in the same account to not have access to a database (not enough permission, private network in a different VPC etc.)

We have touched on the topic of having a per-database health-check several times. I think that will be really beneficial for this scenario, and this could be the best time to add it. How about adding a round of TCP-level connection tests for each dynamic resource prior to the reconciler?

smallinsky commented:
> > @greedy52 I think that this approach is reasonable to reuse the database dynamic `"*": "*"` resource matcher. We can consider checking if db agent that suppose to handle incoming cloud db database discovery resource can actually successfully handle them by ensuring that db agent runs with AWS account by calling [API_GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html) in case of Azure we have similar API like `az account show`. Additionally we can check the AccountID/subscriptionID of an incoming resource and in case of multi AWS/Azure cloud configuration.
> 
> I totally understand the concern. Though in my opinion, I believe this level of filtering is best manually added to the dynamic resource labels. It is possible for a database agent to access databases in a different account (e.g. assume role in a different account for DynamoDB), and it is also possible for a database agent in the same account to not have access to a database (not enough permission, private network in a different VPC etc.)
> 
> We have touched on the topic of having a per-database health-check several times. I think that will be really beneficial for this scenario, and this could be the best time to add it. How about adding a round of TCP-level connection tests for each dynamic resource prior to the reconciler (e.g. during GetCurrentResources)? (it can be quite some work though as we may have to implement some interface for each engine type as some engines have custom URI formats)

This main concerns is that we are recommending to configure db resource by `teleport db configure` command  like: 
```
teleport db configure create \
  -o file \
  --proxy=tele.example.com:443 \
  --token=/tmp/token \
  --azure-postgres-discovery=eastus
  ```
where the default discovery labels are added by default: 
```
db_service:
  enabled: "yes"
  resources:
  - labels:
      "*": "*"
  azure:
  - types: ["postgres"]
    subscriptions:
    - "*"
    resource_groups:
    - "*"
    regions:
    - "eastus"
    tags:
      "*": "*"
 ```


> We have touched on the topic of having a per-database health-check several times. I think that will be really beneficial for this scenario, and this could be the best time to add it.

I think that we can consider adding it but for sure this is a separate feature what needs be discussed independently and have some corner cases.

greedy52 commented:
> This main concerns is that we are recommending to configure db resource by `teleport db configure` command like:
> 
> ```
> teleport db configure create \
>   -o file \
>   --proxy=tele.example.com:443 \
>   --token=/tmp/token \
>   --azure-postgres-discovery=eastus
> ```

How about we change `teleport db configure create` to **NOT** generate any resource labels by default? (or maybe generate ` "teleport.dev/origin": "dynamic"` by default to match current behavior)

Then we can provide extra flags like `--dynamic-resource-labels env=prod,aa=bb` and `--discovery-resource-labels aws-account-id=12345`, that generates something like
```
  resources:
  - labels:
      "teleport.dev/origin": "dynamic"
      "env": "prod"
      "aa": "bb"
  - labels:
      "teleport.dev/origin": "cloud"
      "aws-account-id": "12345"
```
Then we can add a **tip** in the guide for these flags.

It won't help existing database agents but it helps going forward.

greedy52 committed:
give some warnings when discovered database cannot be handled by the database agent

greedy52 committed:
different log levels

greedy52 commented:
@smallinsky this is my first attempt on the warning message. What do you think?

> 2023-01-05T18:04:30Z WARN [DB:SERVIC] Failed to get Azure subscription IDs when checking a database created by the discovery service. You can update "db_service.resources" section of this agent's config file to filter out unwanted resources. If this database is intended to be handled by this agent, please verify that valid cloud credentials are configured for the agent. database:Database(Name=steve-redis-azure, Type=azure, Labels=map[Owner:STeve TeleportDatabaseName:steve-redis-azure engine:Microsoft.Cache/Redis engine-version:6.0 region:Canada Central resource-group:STeve-teleport subscription-id:060a97ea-3a57-4218-9be5-dba3f19ff2b5 teleport.dev/origin:cloud]) error:[
ERROR REPORT:
Original Error: *azidentity.credentialUnavailableError DefaultAzureCredential: failed to acquire a token.
Attempted credentials:
EnvironmentCredential: missing environment variable AZURE_TENANT_ID
ManagedIdentityCredential: IMDS token request timed out
AzureCLICredential: ERROR: Please run &#39;az login&#39; to setup account.
Stack Trace:
github.com/gravitational/teleport/lib/cloud/azure/subscriptions.go:53 github.com/gravitational/teleport/lib/cloud/azure.(*SubscriptionClient).ListSubscriptionIDs
github.com/gravitational/teleport/lib/srv/db/watcher.go:287 github.com/gravitational/teleport/lib/srv/db.(*discoveryResourceCheckerImpl).checkAzure.func1
github.com/gravitational/teleport/lib/utils/fncache.go:135 github.com/gravitational/teleport/lib/utils.FnCacheGet[...].func1
github.com/gravitational/teleport/lib/utils/fncache.go:199 github.com/gravitational/teleport/lib/utils.(*FnCache).get.func1
runtime/asm_amd64.s:1594 runtime.goexit

smallinsky commented:
> @smallinsky this is my first attempt on the warning message. What do you think?
> 
> > 2023-01-05T18:04:30Z WARN [DB:SERVIC] Failed to get Azure subscription IDs when checking a database created by the discovery service. You can update "db_service.resources" section of this agent's config file to filter out unwanted resources. If this database is intended to be handled by this agent, please verify that valid cloud credentials are configured for the agent. database:Database(Name=steve-redis-azure, Type=azure, Labels=map[Owner:STeve TeleportDatabaseName:steve-redis-azure engine:Microsoft.Cache/Redis engine-version:6.0 region:Canada Central resource-group:STeve-teleport subscription-id:060a97ea-3a57-4218-9be5-dba3f19ff2b5 teleport.dev/origin:cloud]) error:[
> > ERROR REPORT:
> > Original Error: *azidentity.credentialUnavailableError DefaultAzureCredential: failed to acquire a token.
> > Attempted credentials:
> > EnvironmentCredential: missing environment variable AZURE_TENANT_ID
> > ManagedIdentityCredential: IMDS token request timed out
> > AzureCLICredential: ERROR: Please run 'az login' to setup account.
> > Stack Trace:
> > github.com/gravitational/teleport/lib/cloud/azure/subscriptions.go:53 github.com/gravitational/teleport/lib/cloud/azure.(*SubscriptionClient).ListSubscriptionIDs
> > github.com/gravitational/teleport/lib/srv/db/watcher.go:287 github.com/gravitational/teleport/lib/srv/db.(*discoveryResourceCheckerImpl).checkAzure.func1
> > github.com/gravitational/teleport/lib/utils/fncache.go:135 github.com/gravitational/teleport/lib/utils.FnCacheGet[...].func1
> > github.com/gravitational/teleport/lib/utils/fncache.go:199 github.com/gravitational/teleport/lib/utils.(*FnCache).get.func1
> > runtime/asm_amd64.s:1594 runtime.goexit

@greedy52 
That looks nice I would suggest removing stack trace from the log because it don't bring any valid context. Also maybe we can provide a link to our docs when we can describe the issue in details.

greedy52 committed:
remove trace and add config reference

greedy52 commented:
> @greedy52 That looks nice I would suggest removing stack trace from the log because it don't bring any valid context. Also maybe we can provide a link to our docs when we can describe the issue in details.

I've changed it to `log.WithField("error", err.Error())` to remove the stack trace. The error itself can be a multi-line string in some cases so they look really bad if embed it directly inside the message so I put it as a field. Also added the reference for the db configuration page, and I will add more info to that doc in a separate PR (planning to add a `--resource-labels` flag to the db configure command too in the same PR).

smallinsky commented on lib/srv/discovery/discovery.go:
```diff
@@ -452,51 +486,50 @@ func (s *Server) initTeleportNodeWatcher() (err error) {
+// splitSlice splits a slice into two, by putting all elements that satisfy the
+// provided check function in the first slice, while putting all other elements
+// in the second slice.
+func splitSlice(s []string, check func(string) bool) (split, other []string) {
```
nit:
```suggestion
func splitSlice(ss []string, check func(string) bool) (split, other []string) {
```

smallinsky commented on lib/srv/discovery/database_watcher.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+func (s *Server) onDatabaseDelete(ctx context.Context, resource types.ResourceWithLabels) error {
+
```
```suggestion
```

smallinsky commented on lib/srv/db/watcher.go:
```diff
@@ -188,13 +199,129 @@ func (s *Server) matcher(resource types.ResourceWithLabels) bool {
+	check(ctx context.Context, database types.Database)
+}
+
+type discoveryResourceCheckerImpl struct {
```
nit: Could we drop the Imp suffix and name it something like `cloudValidCredentailsChecker` ?

smallinsky reviewed (approved):
LGTM.

greedy52 committed:
review comments

GavinFrazar reviewed (approved):
LGTM

greedy52 committed:
add comments and warning log

greedy52 committed:
Merge branch 'master' of github.com:gravitational/teleport into STeve/18742_move_database_discovery_b2

GavinFrazar commented on lib/srv/db/watcher.go:
```diff
@@ -297,14 +299,18 @@ func (c *cloudCredentialsChecker) checkAzure(ctx context.Context, database types
+	}
+
+	if !slices.Contains(allSubIDs, rid.SubscriptionID) {
+		c.warn(nil, database, fmt.Sprintf("The discovered database %q is in a different subscription (ID: %s) than the database agent's tenant.",
```
last minute nit:
```suggestion
		c.warn(nil, database, fmt.Sprintf("The discovered database %q is in a subscription (ID: %s) that the database agent does not have access to.",
```
subscriptions are in tenants, not the other way around

greedy52 committed:
fix UT

greedy52 committed:
Merge branch 'master' of github.com:gravitational/teleport into STeve/18742_move_database_discovery_b2
